segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -600,37 +647,39 @@ public void initialize() throws Exception {
     private CompletableFuture<Void> validateSystemSnapshotExistsInTxn(MetadataTransaction txn, SystemSnapshotRecord systemSnapshot) {
         if (null == systemSnapshot) {
             return CompletableFuture.completedFuture(null);
-        } else {
-            // For each segment in snapshot
-            return Futures.loop(
-                    systemSnapshot.getSegmentSnapshotRecords(),
-                    segmentSnapshot ->
-                        txn.get(segmentSnapshot.segmentMetadata.getKey())
-                            .thenApplyAsync(metadata -> {
-                                Preconditions.checkState(null != metadata);
-                                return metadata;
-                            }, executor)
-                            .thenComposeAsync(v ->
-                                // For each chunk in the segment
-                                Futures.loop(
-                                        segmentSnapshot.getChunkMetadataCollection(),
-                                        m -> txn.get(m.getKey())
-                                                .thenApplyAsync(metadata -> {
-                                                    Preconditions.checkState(null != metadata);
-                                                    return true;
-                                                }, executor),
-                                        executor),
-                            executor)
-                            .thenComposeAsync(vv -> validateSegmentRecordInTxn(txn, segmentSnapshot.segmentMetadata.getKey()), executor)
-                            .thenApplyAsync(v -> true, executor),
-                    executor);
         }
+
+        // For each segment in snapshot
+        return Futures.loop(
+                systemSnapshot.getSegmentSnapshotRecords(),
+                segmentSnapshot ->
+                    txn.get(segmentSnapshot.segmentMetadata.getKey())
+                        .thenApplyAsync(this::checkMetadata, executor)
+                        .thenComposeAsync(v -> validateChunksInSegmentSnapshot(txn, segmentSnapshot), executor)
+                        .thenComposeAsync(vv -> validateSegment(txn, segmentSnapshot.segmentMetadata.getKey()), executor)
+                        .thenApplyAsync(v -> true, executor),
+                executor);
+    }
+
+    private StorageMetadata checkMetadata(StorageMetadata metadata) {
+        Preconditions.checkState(null != metadata);
+        return metadata;
```

Comment:
If you combine the callbacks into fewer, I don't think you need to return `metadata` in this method.

Reply:
Good catch . This method was needed as it has single statement. inlined and removed.